File: packages/dotnet/TraverseEmbedder/TraverseEmbedder/TraverseEmbedder.cs
namespace Traverse.Embedder;

/// <summary>Public .NET surface for Traverse embedder-api/1.0.0.</summary>
public static class TraverseEmbedder
{
    public const string ApiVersion = "1.0.0";
}

public sealed record TraverseBundle(string RootPath, string RuntimeWasmDigest)
{
    public void Validate()
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(RootPath);
        ArgumentException.ThrowIfNullOrWhiteSpace(RuntimeWasmDigest);
    }
}

public sealed record TraverseSubmission(string TargetId, string InputJson)
{
    public void Validate() => ArgumentException.ThrowIfNullOrWhiteSpace(TargetId);
}

public sealed record TraverseSubmissionResult(string SessionId, string Status);

/// <summary>Traceability evidence published with a TraverseEmbedder package release.</summary>
public sealed record TraverseReleaseEvidence(
    string PackageVersion,
    string RuntimeWasmDigest,
    string ConformanceVersion,
    IReadOnlyList<string> SupportedHostVersions)
{
    public void Validate()
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(PackageVersion);
        ArgumentException.ThrowIfNullOrWhiteSpace(RuntimeWasmDigest);
        ArgumentException.ThrowIfNullOrWhiteSpace(ConformanceVersion);
        if (SupportedHostVersions is null || SupportedHostVersions.Count == 0 ||
            SupportedHostVersions.Any(string.IsNullOrWhiteSpace))
        {
            throw new ArgumentException("supported host versions are required", nameof(SupportedHostVersions));
        }
    }
}

/// <summary>Ordered runtime-shaped event exposed by the conformance harness.</summary>
public sealed record TraverseRuntimeEvent(
    int Sequence,
    string TargetId,
    string Status,
    string? InstanceId = null,
    string? EventType = null,
    string? SessionId = null,
    string? ErrorData = null,
    string? Output = null);

public sealed record TraverseCompatibleResult(string? InstanceId, string Status);

/// <summary>
/// Deterministic conformance test double. It never evaluates application
/// business logic and never starts a Traverse sidecar process.
/// </summary>
public sealed class InMemoryTraverseEmbedder
{
    private TraverseBundle? bundle;
    private int submissionSequence;
    private int compatibleSequence;
    private readonly List<TraverseRuntimeEvent> events = [];
    private readonly Dictionary<string, string> compatibleInstances = [];
    private string? targetOutput;

    public InMemoryTraverseEmbedder WithTargetOutput(string output)
    {
        targetOutput = output;
        return this;
    }

    public void Initialize(TraverseBundle value)
    {
        if (bundle is not null) throw new InvalidOperationException("embedder is already initialized");
        value.Validate();
        bundle = value;
    }

    public void Shutdown()
    {
        bundle = null;
        submissionSequence = 0;
        compatibleSequence = 0;
        events.Clear();
        compatibleInstances.Clear();
    }

    public TraverseSubmissionResult Submit(TraverseSubmission submission)
    {
        if (bundle is null) throw new InvalidOperationException("embedder is not initialized");
        submission.Validate();
        submissionSequence++;
        var result = new TraverseSubmissionResult($"dotnet-session-{submissionSequence}", "accepted");
        events.Add(new TraverseRuntimeEvent(submissionSequence, submission.TargetId, result.Status,
            EventType: targetOutput is null ? null : "capability_result",
            SessionId: targetOutput is null ? null : result.SessionId, Output: targetOutput));
        return result;
    }

    /// <summary>Returns ordered runtime-shaped events emitted after a sequence cursor.</summary>
    public IReadOnlyList<TraverseRuntimeEvent> Subscribe(int afterSequence = 0)
    {
        EnsureInitialized();
        return events.Where(@event => @event.Sequence > afterSequence).ToArray();
    }

    public TraverseCompatibleResult CompatibleStart(string capabilityId, string inputJson)
    {
        EnsureInitialized();
        ArgumentException.ThrowIfNullOrWhiteSpace(capabilityId);
        compatibleSequence++;
        var instanceId = $"dotnet-compatible-{compatibleSequence}";
        compatibleInstances[capabilityId] = instanceId;
        AppendCompatibleEvent(capabilityId, instanceId, "started");
        return new TraverseCompatibleResult(instanceId, "started");
    }

    public TraverseCompatibleResult CompatibleStop(string capabilityId, string? instanceId)
    {
        var resolvedInstanceId = CompatibleInstance(capabilityId, instanceId);
        compatibleInstances.Remove(capabilityId);
        AppendCompatibleEvent(capabilityId, resolvedInstanceId, "stopped");
        return new TraverseCompatibleResult(resolvedInstanceId, "stopped");
    }

    public TraverseCompatibleResult CompatibleKill(string capabilityId, string? instanceId)
    {
        var resolvedInstanceId = CompatibleInstance(capabilityId, instanceId);
        compatibleInstances.Remove(capabilityId);
        AppendCompatibleEvent(capabilityId, resolvedInstanceId, "killed");
        return new TraverseCompatibleResult(resolvedInstanceId, "killed");
    }

    private string CompatibleInstance(string capabilityId, string? instanceId)
    {
        EnsureInitialized();
        if (!compatibleInstances.TryGetValue(capabilityId, out var activeInstanceId) ||
            (instanceId is not null && instanceId != activeInstanceId))
        {
            throw new InvalidOperationException("compatible instance is not active");
        }
        return activeInstanceId;
    }

    private void AppendCompatibleEvent(string capabilityId, string instanceId, string status)
    {
        submissionSequence++;
        events.Add(new TraverseRuntimeEvent(submissionSequence, capabilityId, status, instanceId));
    }

    private void EnsureInitialized()
    {
        if (bundle is null) throw new InvalidOperationException("embedder is not initialized");
    }
}

File: packages/kotlin/TraverseEmbedder/traverse-embedder/src/main/kotlin/dev/traverse/embedder/TraverseEmbedder.kt
package dev.traverse.embedder

/** Public Kotlin/Android surface for Traverse `embedder-api/1.0.0`. */
object TraverseEmbedder {
    const val API_VERSION = "1.0.0"
}

data class TraverseBundle(val rootPath: String, val runtimeWasmDigest: String) {
    init {
        require(rootPath.isNotBlank()) { "bundle root path is required" }
        require(runtimeWasmDigest.isNotBlank()) { "runtime WASM digest is required" }
    }
}

data class TraverseSubmission(val targetId: String, val inputJson: String) {
    init { require(targetId.isNotBlank()) { "target_id is required" } }
}

data class TraverseSubmissionResult(val sessionId: String, val status: String)

/** Traceability evidence published with a TraverseEmbedder package release. */
data class TraverseReleaseEvidence(
    val packageVersion: String,
    val runtimeWasmDigest: String,
    val conformanceVersion: String,
    val supportedHostVersions: List<String>,
) {
    init {
        require(packageVersion.isNotBlank()) { "package version is required" }
        require(runtimeWasmDigest.isNotBlank()) { "runtime WASM digest is required" }
        require(conformanceVersion.isNotBlank()) { "conformance version is required" }
        require(supportedHostVersions.isNotEmpty() && supportedHostVersions.all { it.isNotBlank() }) {
            "supported host versions are required"
        }
    }
}

/** Ordered runtime-shaped event exposed by the deterministic conformance harness. */
data class TraverseRuntimeEvent(
    val sequence: Int,
    val targetId: String,
    val status: String,
    val instanceId: String? = null,
    val eventType: String? = null,
    val sessionId: String? = null,
    val errorData: String? = null,
    val output: String? = null,
)

data class TraverseCompatibleResult(val instanceId: String?, val status: String)

/** Deterministic test double; it never evaluates application business logic. */
class InMemoryTraverseEmbedder {
    private var bundle: TraverseBundle? = null
    private var submissionSequence = 0
    private var compatibleSequence = 0
    private val events = mutableListOf<TraverseRuntimeEvent>()
    private val compatibleInstances = mutableMapOf<String, String>()
    private var targetOutput: String? = null

    fun withTargetOutput(output: String): InMemoryTraverseEmbedder = apply { targetOutput = output }

    fun initialize(bundle: TraverseBundle) {
        check(this.bundle == null) { "embedder is already initialized" }
        this.bundle = bundle
    }

    fun shutdown() {
        bundle = null
        submissionSequence = 0
        compatibleSequence = 0
        events.clear()
        compatibleInstances.clear()
    }

    fun submit(submission: TraverseSubmission): TraverseSubmissionResult {
        check(bundle != null) { "embedder is not initialized" }
        submissionSequence += 1
        val result = TraverseSubmissionResult("kotlin-session-$submissionSequence", "accepted")
        events += TraverseRuntimeEvent(submissionSequence, submission.targetId, result.status, eventType = if (targetOutput == null) null else "capability_result", sessionId = if (targetOutput == null) null else result.sessionId, output = targetOutput)
        return result
    }

    /** Returns ordered runtime-shaped events emitted after the supplied cursor. */
    fun subscribe(afterSequence: Int = 0): List<TraverseRuntimeEvent> {
        check(bundle != null) { "embedder is not initialized" }
        return events.filter { it.sequence > afterSequence }
    }

    fun compatibleStart(capabilityId: String, inputJson: String): TraverseCompatibleResult {
        check(bundle != null) { "embedder is not initialized" }
        require(capabilityId.isNotBlank()) { "capability_id is required" }
        compatibleSequence += 1
        val instanceId = "kotlin-compatible-$compatibleSequence"
        compatibleInstances[capabilityId] = instanceId
        appendCompatibleEvent(capabilityId, instanceId, "started")
        return TraverseCompatibleResult(instanceId, "started")
    }

    fun compatibleStop(capabilityId: String, instanceId: String?): TraverseCompatibleResult {
        val resolvedInstanceId = compatibleInstance(capabilityId, instanceId)
        compatibleInstances.remove(capabilityId)
        appendCompatibleEvent(capabilityId, resolvedInstanceId, "stopped")
        return TraverseCompatibleResult(resolvedInstanceId, "stopped")
    }

    fun compatibleKill(capabilityId: String, instanceId: String?): TraverseCompatibleResult {
        val resolvedInstanceId = compatibleInstance(capabilityId, instanceId)
        compatibleInstances.remove(capabilityId)
        appendCompatibleEvent(capabilityId, resolvedInstanceId, "killed")
        return TraverseCompatibleResult(resolvedInstanceId, "killed")
    }

    private fun compatibleInstance(capabilityId: String, instanceId: String?): String {
        check(bundle != null) { "embedder is not initialized" }
        val activeInstanceId = compatibleInstances[capabilityId]
        check(activeInstanceId != null && (instanceId == null || instanceId == activeInstanceId)) {
            "compatible instance is not active"
        }
        return activeInstanceId
    }

    private fun appendCompatibleEvent(capabilityId: String, instanceId: String, status: String) {
        submissionSequence += 1
        events += TraverseRuntimeEvent(submissionSequence, capabilityId, status, instanceId)
    }
}
